pubsub/responses.go

```diff
@@ -13,19 +13,86 @@ limitations under the License.
+type BulkPublishStatus string
+
 const (
 	// Success means the message is received and processed correctly.
 	Success AppResponseStatus = "SUCCESS"
 	// Retry means the message is received but could not be processed and must be retried.
 	Retry AppResponseStatus = "RETRY"
 	// Drop means the message is received but should not be processed.
 	Drop AppResponseStatus = "DROP"
+	// PublishSucceeded represents that message was published successfully.
+	PublishSucceeded BulkPublishStatus = "SUCCESS"
+	// PublishFailed represents that message publishing Failed.
+	PublishFailed BulkPublishStatus = "FAILED"
 )
 
 // AppResponse is the object describing the response from user code after a pubsub event.
 type AppResponse struct {
 	Status AppResponseStatus `json:"status"`
 }
+
+// AppBulkResponseEntry Represents single response, as part of AppBulkResponse, to be
+// sent by subscibed App for the corresponding single message during bulk subscribe
+type AppBulkResponseEntry struct {
+	EntryID string            `json:"entryID"`
+	Status  AppResponseStatus `json:"status"`
+}
+
+// AppBulkResponse is the whole bulk subscribe response sent by App
+type AppBulkResponse struct {
+	AppResponses []AppBulkResponseEntry `json:"statuses"`
+}
+
+// BulkPublishResponseEntry Represents single publish response, as part of BulkPublishResponse
+// to be sent to publishing App for the corresponding single message during bulk publish
+type BulkPublishResponseEntry struct {
+	EntryID string            `json:"entryID"`
+	Status  BulkPublishStatus `json:"status"`
+	Error   error             `json:"error"`
+}
+
+// BulkPublishResponse is the whole bulk publish response sent to App
+type BulkPublishResponse struct {
+	Statuses []BulkPublishResponseEntry `json:"statuses"`
+}
+
+// BulkSubscribeResponseEntry Represents single subscribe response item, as part of BulkSubscribeResponse
+// to be sent to building block for the corresponding single message during bulk subscribe
+type BulkSubscribeResponseEntry struct {
+	EntryID string `json:"entryID"`
+	Error   string `json:"error"`
+}
+
+// BulkSubscribeResponse is the whole bulk subscribe response sent to building block
+type BulkSubscribeResponse struct {
+	Error    string                       `json:"error"`
+	Statuses []BulkSubscribeResponseEntry `json:"statuses"`
+}
+
+// NewBulkPublishResponse returns a BulkPublishResponse with each entry having same status and error.
+// This method is a helper method to map a single error/success response on BulkPublish to multiple events.
+func NewBulkPublishResponse(messages []BulkMessageEntry, status BulkPublishStatus, err error) BulkPublishResponse {
+	response := BulkPublishResponse{}
+	for i, msg := range messages {
+		st := BulkPublishResponseEntry{}
+		if msg.EntryID != "" {
+			st.EntryID = msg.EntryID
+		} else {
+			msg.EntryID = fmt.Sprintf("%d", i)
+		}
+		st.Status = status
+		if err != nil {
+			st.Error = err
+		}
+		response.Statuses = append(response.Statuses, st)
```

Comment:
Could this array be pre-allocated, since we know the number of statuses? This way we can skip the `append` and reduce the number of allocations in a hot path.